### ISSUE TITLE
Product Variation Ordering Not Persisting After Drag-and-Drop

### ISSUE THREAD

**Issue by nicdwilson** [woothemes/woocommerce#63332]
#### A clear and concise description of what the bug is.

When manually adding new product variations and immediately reordering them via drag-and-drop without an intermediate save, the variations do not save in the expected order. Multiple variations can end up with duplicate `menu_order` values, causing them to appear in unexpected positions.

**Reproduction Rate**

* Intermittent when only one new variation is added
* Consistent when multiple variations are added in sequence without intermediate saves
* Does not occur if you save immediately after creating each variation

#### Write steps here

1. Create or edit a variable product with at least 3 existing saved variations
2. Click "Add manually" to add a new variation
3. Fill in the variation details (attributes, price, etc.)
4. **Without saving**, drag the new variation to a specific position (e.g., to the bottom of the list)
5. Click "Save changes"
6. Observe the variation order after page refresh

**The Problem Sequence**

1. New variations are created with `menu_order = 0` (default for new posts)
   * File: `includes/class-wc-ajax.php:850-869` (`add_variation()`)
   * No explicit `menu_order` is set during creation
2. User drags variations to reorder (JavaScript updates hidden inputs)
   * File: `assets/js/admin/meta-boxes-product-variation.js:439-464`
   * The `variation_row_indexes()` function updates `variation_menu_order` input fields
   * These values are sent in POST data when user clicks "Save"
3. On save, `sort_all_product_variations()` runs FIRST (line 474)

* File: `includes/data-stores/class-wc-product-data-store-cpt.php:1550-1566`
* Queries all variations: `ORDER BY menu_order ASC, ID ASC`
* New variation with `menu_order = 0` appears first in results
* Re-assigns menu_order sequentially (1, 2, 3, 4, ...) based on this stale order
* This overwrites the user's intended drag-drop order

4. Then individual variation saves process with POST data
   * But the database was already modified in step 3
   * Results in conflicts and duplicate `menu_order` values

**Evidence from Reproduction Testing**

Test logs show the exact problem:

```
[NEW VARIATION CREATED] Variation ID: 125 | Initial menu_order: 0
(User drags to position 5 - last position)
[AFTER SAVE] Variation ID: 124 | Final DB menu_order: 5
[AFTER SAVE] Variation ID: 125 | Final DB menu_order: 5  ← DUPLICATE!
```

**Result:** Both variations 124 and 125 have `menu_order = 5`. When displayed with `ORDER BY menu_order ASC, ID ASC`, variation 124 (lower ID) appears last, and variation 125 appears "one up" from the bottom.

#### A clear and concise description of what you expected to happen.

New variation appears in the position where it was dragged (bottom of list)

## Environment

Please fill out any of the below that applies

**Comment by jamesckemp** [comment]
Thanks for the detailed investigation, the code references and test logs are really helpful. I'll get this to our product team to investigate further. They'll keep you updated here.

**Comment by billbobong** [comment]
Since a couple of months ago I have the same problem on every new Woo product created. In Woo / Attributes, I already have “Color” listed with many colors in a particular sequence. And in “Size” I have 8 sizes in sequence from S-5XL. And in each new product in Woo / Products / [Product] / Product Data / Attributes I have ensured “Color” with colors in the same sequence again and “Size” with 8 sizes in the same sequence again. Then in Variations, now when I hit “Generate Variations”, all the variations are listed in alpha-numeric order and completely ignore my custom order (yes, I have selected “Custom ordering”). Previously some months ago, hitting “Generate Variations” obeyed the correct order, I also get the same problem with other plugins deactivated and themes changed. 

I need all my data organised and in the correct order and jumbling it up, like re-shuffling a pack of cards, is EXTREMELY time consuming to rectify and very DAMAGING to my business in lost time and revenue. So could Sunyatasattva please please PLEASE treat this as URGENT and get it fixed without further delay?
